BookNLPcharTracker.ipynb

```diff
@@ -2097,10 +40,61 @@
         "        }\n",
         "        self.booknlp = BookNLP(\"en\", self.model_params)\n",
         "\n",
+        "    def preprocess_text(self, input_file, output_file):\n",
+        "        \"\"\"\n",
+        "        Preprocess the book text by removing the introduction, table of contents, ending sections,\n",
+        "        text with leading spaces, and text enclosed in square brackets or between underscores.\n",
+        "        Removes unwanted punctuation but keeps sentence-ending punctuation marks.\n",
+        "        \"\"\"\n",
+        "        with open(input_file, 'r', encoding='utf-8') as f:\n",
+        "            text = f.read()\n",
+        "\n",
+        "        text = re.sub(r'_.*?_', '', text)\n",
+        "\n",
+        "        text = re.sub(r'\\[.*?\\]', '', text, flags=re.DOTALL)\n",
+        "\n",
+        "        text = re.sub(r\"[\\\"'“”‘’\\]\\[\\(\\){}]\", '', text)\n",
+        "\n",
+        "        text = '\\n'.join([line for line in text.splitlines() if not re.match(r'^\\s{2,}', line)])\n",
```

Comment:
deleting lines starting with spaces is also tricky because there are some letters at the end of the book that have spaces at the start

Reply:
I see the issue, it was recognizing the letters inside the text as introduction/ending. It should recognize extra tabulators instead. I've added the print statement, so we can see what's going on.